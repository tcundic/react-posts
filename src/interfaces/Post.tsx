
interface Post {
    id: number,
    title: string,
    content: string,
    user: string,
    comments: Comment[]
};

interface Comment {
    id: number,
    content: string
}

export default Post;