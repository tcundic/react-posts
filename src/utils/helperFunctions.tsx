export const greet = (greeting: string, componentName: string): void => console.log(`${greeting} ${componentName}`);