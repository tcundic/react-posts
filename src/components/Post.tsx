import React, {useEffect, useState} from "react";
import {greet} from "../utils/helperFunctions";
import {useParams} from "react-router-dom";
import {getPostById} from "../utils/APIUtils";
import IPost from "../interfaces/Post";

const Post = ({greeting}: {greeting: string}) => {
    const { id }: {id: string} = useParams();
    const [post, setPost] = useState<IPost>();

    useEffect(() => greet(greeting, Post.name), [greeting]);

    useEffect(() => {
        getPostById(parseInt(id)).then((post: IPost) => {
            setPost(post);
        });
    }, [id]);

    return (
        <div className="mt-6 mx-6">
            <div className="post-header">
                <h1>{post?.title}</h1>
                <div className="post-username-caption padding-x">{post?.user && `@${post?.user}`}</div>
            </div>
            <div className="post-content mt-6">{post?.content}</div>
            {post?.comments.map(comment => (
                <div className="post-comment padding-x" key={`${post.id}-${comment.id}`}>
                    {comment.content}
                </div>
            ))}
        </div>
    );
};

export default Post;