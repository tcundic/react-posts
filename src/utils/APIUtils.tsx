import {apiEndpoint} from "../config";
import RequestOptions from "../interfaces/Request";
import Post from "../interfaces/Post";

const request = (options: RequestOptions, additionalHeaders?: Object): Promise<any> => {
    const headers = {
        'Content-Type': 'application/json',
        ...additionalHeaders
    };

    const defaults = {headers};
    options = Object.assign({}, defaults, options, {crossDomain: true});

    return fetch(options.url, options)
        .then(response => response.json());
};

export function getPosts(): Promise<Post[]> {
    return request({
        url: `${apiEndpoint}posts`,
        method: 'GET'
    });
}

export function getPostById(id: number): Promise<Post> {
    return request({
        url: `${apiEndpoint}posts/${id}`,
        method: 'GET'
    });
}
