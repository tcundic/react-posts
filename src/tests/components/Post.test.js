import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as Greet from "../../utils/helperFunctions";
import React from "react";
import posts from "../fixtures/posts";
import { act } from "@testing-library/react";
import Post from "../../components/Post";
import { BrowserRouter, MemoryRouter } from 'react-router-dom';

beforeEach(() => {
    jest.spyOn(global, 'fetch').mockResolvedValue({
        json: jest.fn().mockResolvedValue(posts[0])
    })
});

afterEach(() => {
    jest.restoreAllMocks();
});

const waitForComponentToPaint = async (wrapper) => {
    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
        wrapper.update();
    });
};

const greeting = "Hello from";
const name = Post.name;

test('should render Posts correctly', () => {
    const wrapper = shallow(
        <MemoryRouter initialEntries={['/post/1']}>
            <Post greeting={greeting}/>
        </MemoryRouter>);
    expect(toJson(wrapper)).toMatchSnapshot();
});

test('should print Posts greeting', () => {
    const greetSpy = jest.spyOn(Greet, 'greet');
    const wrapper = mount(
        <MemoryRouter initialEntries={['/post/1']}>
            <Post greeting={greeting}/>
        </MemoryRouter>);
    waitForComponentToPaint(wrapper);
    expect(greetSpy).toHaveBeenLastCalledWith(greeting, name);
});
