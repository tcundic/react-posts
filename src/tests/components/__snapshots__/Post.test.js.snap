// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Posts correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "5ygdbm",
          "pathname": "/post/1",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "5ygdbm",
        "pathname": "/post/1",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Post
    greeting="Hello from"
  />
</Router>
`;
